internal/jimm/bootstrap/bootstrap.go

```diff
@@ -362,42 +395,47 @@ func (b *bootstrapManager) runBootstrap(
 	}
 	defer cleanup()
 
-	for output := range outputCh {
-		if output.Err != nil {
-			if writeLogErr := b.store.AddBootstrapLog(
-				ctx,
-				jobId,
-				output.Err.Error(),
-			); writeLogErr != nil {
-				zapctx.Error(ctx, "failed to write bootstrap error", zap.Error(writeLogErr), zap.String("jobId", jobId.String()))
-			}
-			return errors.E(fmt.Errorf("bootstrap command failed: %w", output.Err))
+	err = b.consumeCommandOutput(ctx, outputCh, jobId)
+	if err != nil {
+		return err
+	}
+
+	// controllerCleanup is a helper function to cleanup the controller if we fail at any point
+	// after the bootstrap, avoiding orphaned controllers in the cloud.
+	controllerCleanup := func(err error, controllerDetails *jujuclient.ControllerDetails) error {
```

Comment:
This is a closure because it felt more appropriate than a separate method.
I could also use `defer func()` but then I would need to make the error return value a named parameter in order to access it from the deferred function (I think?) so this felt simpler but can definitely change to `defer ...` if that's preferred.

Reply:
This seems fine to me honestly. I think I would prefer it in its own function but it is ok to leave it.